cqrs-postgres/src/store.rs

```diff
@@ -287,6 +288,107 @@ where
         ])?;
         Ok(rows.iter().map(|r| r.get(0)).collect())
     }
+
+    /// Reads events and associated metadata from the event source for a given identifier.
+    ///
+    /// Only loads events after the event number provided in `since` (See [Since]), and will only load a maximum of
+    /// `max_count` events, if given. If not given, will read all remaining events.
+    pub fn read_events_with_metadata<I>(
+        &self,
+        id: &I,
+        since: Since,
+        max_count: Option<u64>,
+    ) -> Result<
+        Option<Vec<Result<VersionedEventWithMetadata<E, M>, LoadError<E::Error>>>>,
+        LoadError<E::Error>,
+    >
+    where
+        I: AggregateId<A>,
+        E: DeserializableEvent,
+        M: for<'de> serde::Deserialize<'de>,
+    {
+        let last_sequence = match since {
+            cqrs_core::Since::BeginningOfStream => 0,
+            cqrs_core::Since::Event(x) => x.get(),
+        } as i64;
+
+        let events;
+        let trans = self
+            .conn
+            .transaction_with(postgres::transaction::Config::default().read_only(true))?;
+
+        let handle_row = |row: postgres::rows::Row| {
+            let sequence: Sequence = row.get(0);
+            let event_type: String = row.get(1);
+            let raw: RawJsonRead = row.get(2);
+            let metadata: Json<M> = row.get(3);
+            let event = E::deserialize_event_from_buffer(&raw.0, &event_type)
+                .map_err(LoadError::DeserializationError)?
+                .ok_or_else(|| LoadError::UnknownEventType(event_type.clone()))?;
+            log::trace!(
+                "entity {}: loaded event; sequence: {}, type: {}",
+                id.as_str(),
+                sequence.0,
+                event_type
+            );
+            Ok(VersionedEventWithMetadata {
+                sequence: sequence.0,
+                event,
+                metadata: metadata.0,
+            })
+        };
+
+        let stmt;
+        {
+            let mut rows;
+            if let Some(max_count) = max_count {
+                stmt = trans.prepare_cached(
+                    "SELECT sequence, event_type, payload, metadata \
+                     FROM events \
+                     WHERE aggregate_type = $1 AND entity_id = $2 AND sequence > $3 \
+                     ORDER BY sequence ASC \
+                     LIMIT $4",
+                )?;
+                rows = stmt.lazy_query(
+                    &trans,
+                    &[
+                        &A::aggregate_type(),
+                        &id.as_str(),
+                        &last_sequence,
+                        &(max_count.min(i64::max_value() as u64) as i64),
+                    ],
+                    100,
```

Comment:
Should the row_limit be hard capped at 100 here and on line 372?

Reply:
That's a maximum of 100 rows per lazy load. If we have a `max_count` of `Some(1000)` and an event stream with that many events, then we will load all 1000 events, but internally we will grab 100 at a time rather than requiring the server to stream all 1000 over at once. In general, at our loads, we're probably loading fewer rows at a time, but if we have really long histories and ask for all the events, we would put a heavy burden on the server to serve all of our request first.